File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/Tester.java
package org.firstinspires.ftc.teamcode.TeleOp;

import static java.lang.Math.PI;

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
import com.qualcomm.robotcore.hardware.DcMotorEx;
import com.qualcomm.robotcore.hardware.Servo;
import com.qualcomm.robotcore.util.ElapsedTime;

/*
Tester, this is where we write and test all of our code,
ALWAYS test and write new code here, no where else
 */

@TeleOp(name="Tester", group="Controlled")
public class Tester extends LinearOpMode {
    ElapsedTime runtime = new ElapsedTime();
    //pre init
    //motors
    DcMotorEx FLmotor;
    DcMotorEx FRmotor;
    DcMotorEx BLmotor;
    DcMotorEx BRmotor;
    DcMotorEx Xrail;

    //Servos
    Servo gripperL;
    Servo gripperR;

    //speed variables
    double TurnSpeed = .5;
    double drivespeed = .7;

    //servo positions

    public void runOpMode() {

        //getting motors
        BLmotor = hardwareMap.get(DcMotorEx.class, "BLmotor");
        FLmotor = hardwareMap.get(DcMotorEx.class, "FLmotor");
        BRmotor = hardwareMap.get(DcMotorEx.class, "BRmotor");
        FRmotor = hardwareMap.get(DcMotorEx.class, "FRmotor");
        Xrail = hardwareMap.get(DcMotorEx.class, "Xrail");



        BRmotor.setZeroPowerBehavior(DcMotorEx.ZeroPowerBehavior.BRAKE);
        FLmotor.setZeroPowerBehavior(DcMotorEx.ZeroPowerBehavior.BRAKE);
        BRmotor.setZeroPowerBehavior(DcMotorEx.ZeroPowerBehavior.BRAKE);
        FRmotor.setZeroPowerBehavior(DcMotorEx.ZeroPowerBehavior.BRAKE);
        Xrail.setZeroPowerBehavior(DcMotorEx.ZeroPowerBehavior.BRAKE);

        //motor directions
        BLmotor.setDirection(DcMotorEx.Direction.REVERSE);
        FLmotor.setDirection(DcMotorEx.Direction.REVERSE);
        BRmotor.setDirection(DcMotorEx.Direction.REVERSE);
        FRmotor.setDirection(DcMotorEx.Direction.REVERSE);
        Xrail.setDirection(DcMotorEx.Direction.REVERSE);

        //getting servos
        gripperL = hardwareMap.get(Servo.class,"gripperL");
        gripperR = hardwareMap.get(Servo.class,"gripperR");

        //servo directions
        gripperR.setDirection(Servo.Direction.FORWARD);
        gripperL.setDirection(Servo.Direction.REVERSE);

        //init
        openGrippers();

        waitForStart();
        runtime.reset();

        while (opModeIsActive()) {

            if(gamepad2.a){
                closeGrippers();
            }
            else{
                openGrippers();
            }

            if(gamepad2.b){
                XrailMove(.3,5);
            }

            if(gamepad2.y){
                XrailMove(.3,5);
            }
            if(gamepad2.x){
                XrailMove(.3,5);
            }

            //joystick Xrail
            Xrail.setPower(gamepad2.left_stick_y * .8);

            //drive
            BLmotor.setPower(((gamepad1.right_trigger - gamepad1.left_trigger) * TurnSpeed) + (gamepad1.left_stick_y + (-gamepad1.right_stick_x))*(drivespeed));
            FLmotor.setPower(((gamepad1.right_trigger - gamepad1.left_trigger) * TurnSpeed) + (gamepad1.left_stick_y - (-gamepad1.right_stick_x))*(drivespeed));
            BRmotor.setPower(((gamepad1.right_trigger - gamepad1.left_trigger) * TurnSpeed) - (gamepad1.left_stick_y - (-gamepad1.right_stick_x))*(drivespeed));
            FRmotor.setPower(((gamepad1.right_trigger - gamepad1.left_trigger) * TurnSpeed) - (gamepad1.left_stick_y + (-gamepad1.right_stick_x))*(drivespeed));
            //end of drive
        }

    }
    public void XrailMove(double power, double Inches) {
        double TicksPerRevolution = 537.7;
        // diameter Millimeters
        double DiameterMM = 38.1;
        // diameter Inches
        double DiameterIn = DiameterMM / 25.4; //3.7795275590551185 Inches
        // Circumference
        double Circumference = DiameterIn * PI; //11.873736013567724 Inches
        // Finding ticks per inch
        double TicksPerInch = TicksPerRevolution / Circumference; //45.28482015985433 Inches

        int distance = (int) (Inches * TicksPerInch);

        Xrail.setMode(DcMotorEx.RunMode.STOP_AND_RESET_ENCODER); //resetting the encoder to clear it
        Xrail.setTargetPosition(distance); //giving the encoder a target position
        Xrail.setMode(DcMotorEx.RunMode.RUN_TO_POSITION); //telling it to go to that position
        Xrail.setPower(power); //giving it the power

        while (Xrail.isBusy()) {
        }
        Xrail.setPower(0); //just for when its done
        Xrail.setMode(DcMotorEx.RunMode.RUN_USING_ENCODER);
    }
    public void closeGrippers(){
        gripperL.setPosition(.1);
        gripperR.setPosition(.1);
    }
    public void openGrippers(){
        gripperL.setPosition(.9);
        gripperR.setPosition(.9);
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/Kotlin/Tester.kt
package org.firstinspires.ftc.teamcode.TeleOp.Kotlin

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import com.qualcomm.robotcore.eventloop.opmode.OpMode
import com.qualcomm.robotcore.eventloop.opmode.TeleOp
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.Servo
import com.qualcomm.robotcore.util.ElapsedTime

/*
Tester, this is where we write and test all of our code,
ALWAYS test and write new code here, no where else
 */

@TeleOp(name = "TesterKt", group = "Controlled")
class Tester : LinearOpMode() {
    var runtime = ElapsedTime()

    //pre init
    //motors
    var BLmotor: DcMotorEx? = null
    var FLmotor: DcMotorEx? = null
    var BRmotor: DcMotorEx? = null
    var FRmotor: DcMotorEx? = null
    var XrailL: DcMotorEx? = null
    var XrailR: DcMotorEx? = null
    var longGripper: DcMotorEx? = null

    //Servos
    var gripperL: Servo? = null
    var gripperR: Servo? = null
    var left: Servo? = null
    var middle: Servo? = null
    var right: Servo? = null


    //speed variables
    var TurnSpeed = .7
    var drivespeed = .7

    override fun runOpMode() {
        //getting motors
        BLmotor = hardwareMap.get(DcMotorEx::class.java, "BLmotor")
        FLmotor = hardwareMap.get(DcMotorEx::class.java, "FLmotor")
        BRmotor = hardwareMap.get(DcMotorEx::class.java, "BRmotor")
        FRmotor = hardwareMap.get(DcMotorEx::class.java, "FRmotor")
        XrailL = hardwareMap.get(DcMotorEx::class.java, "XrailL")
        XrailR = hardwareMap.get(DcMotorEx::class.java, "XrailR")
        longGripper = hardwareMap.get(DcMotorEx::class.java, "longGripper")

        //making them brake
        BLmotor?.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        FLmotor?.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        BRmotor?.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        FRmotor?.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        XrailL?.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        XrailR?.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        longGripper?.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE

        //motor directions
        BLmotor?.direction = DcMotorSimple.Direction.REVERSE
        FLmotor?.direction = DcMotorSimple.Direction.REVERSE
        BRmotor?.direction = DcMotorSimple.Direction.REVERSE
        FRmotor?.direction = DcMotorSimple.Direction.REVERSE
        XrailL?.direction = DcMotorSimple.Direction.FORWARD
        XrailR?.direction = DcMotorSimple.Direction.FORWARD
        longGripper?.direction = DcMotorSimple.Direction.FORWARD

        //getting servos
        gripperL = hardwareMap.get(Servo::class.java, "gripperL")
        gripperR = hardwareMap.get(Servo::class.java, "gripperR")
        left = hardwareMap.get(Servo::class.java, "left")
        middle = hardwareMap.get(Servo::class.java, "middle")
        right = hardwareMap.get(Servo::class.java, "right")


        //servo directions

        //servo directions
        gripperR?.direction = Servo.Direction.FORWARD
        gripperL?.direction = Servo.Direction.REVERSE
        left?.direction = Servo.Direction.FORWARD
        middle?.direction = Servo.Direction.FORWARD
        right?.direction = Servo.Direction.FORWARD



        //init

        //servo positions
        gripperL?.position = (.9)
        gripperR?.position = (.5)
        left?.position = (.5)
        middle?.position = (.5)
        right?.position = (.5)


        runtime.reset()

        if (gamepad2.a) {
            gripperL?.position = .1
            gripperL?.position = .1
            telemetry.addLine("gripping")
            telemetry.update()
        } else {
            gripperL?.position = .9
            gripperL?.position = .5
            telemetry.addLine("not gripping :( ")
            telemetry.update()
        }

        //move the slides
        XrailL?.power = ((gamepad2.left_stick_y * 1).toDouble())
        XrailR?.power = ((gamepad2.left_stick_y * 1).toDouble())

        //move the big rail thingy
        longGripper?.power = ((gamepad2.right_stick_y * 1).toDouble())

        //drive
        BLmotor?.power = ((gamepad1.right_trigger - gamepad1.left_trigger) * TurnSpeed + (gamepad1.left_stick_y + gamepad1.right_stick_x) * drivespeed)
        FLmotor?.power = ((gamepad1.right_trigger - gamepad1.left_trigger) * TurnSpeed + (gamepad1.left_stick_y - gamepad1.right_stick_x) * drivespeed)
        BRmotor?.power = ((gamepad1.right_trigger - gamepad1.left_trigger) * TurnSpeed - (gamepad1.left_stick_y - gamepad1.right_stick_x) * drivespeed)
        FRmotor?.power = ((gamepad1.right_trigger - gamepad1.left_trigger) * TurnSpeed - (gamepad1.left_stick_y + gamepad1.right_stick_x) * drivespeed)
    }
}



